### PR TITLE
replace usages of md5 with blake3

### PR DESCRIPTION
**Pull request by benjones**
Add an implementation of Blake3 cryptographic hash function to replace MD5.  The default hash length for blake3 is 32 bytes rather than 16, so I use a truncated version of the hash to keep the lengths the same as they were with MD5.  Since Blake3 is designed to produces hashes of any desired length, I'm assuming that a shorter than default hash is at least a collision resistant as md5 (though I didn't find a guarantee of this)

The implementation here is a bit simplified from the reference implementation because it doesn't support streaming (the API is just `ubyte[32] blake3(ubyte[])`).

I'm assuming there are consequences of this change that I'm not aware of and hence not fixing here, so LMK what else needs to be fixed.

The rationale for this change is that MD5 is considered insecure with known weaknesses.  While these are probably not exploitable in the compiler, I think it's a good PR move to dump MD5.  Also, blake3 is supposed to be faster than MD5, though I haven't benchmarked this change (and some of the speedups come from being able to parallelize hashing big inputs which won't be a concern in the compiler).

Also, I'm on an ARM Mac so can't test this locally (yet, I'll work that early next week).

I assume I also need to add some copyright stuff, etc.

### CODE REVIEW

**Comment by benjones** [comment]
> It would be nice to run a benchmark by compiling s huge project, to see the benefits

Agree, I'll try to run something on a linux box sometime this week

**Comment by benjones** [comment]
I don't understand this compiler error in the CI stuff?  It builds/runs fine using ldc locally.  Where is there an attempted implicit cast to uint?

```

 ulong cursor = 0;

 //greater because if it's == we still need to finalize the last
 //chunk
 while(data.length - cursor > ChunkLength){

     const ubyte[] chunk = data[cursor .. cursor + ChunkLength];
```

```
src/dmd/common/blake3.d(34): Error: cannot implicitly convert expression `cursor` of type `ulong` to `uint`
src/dmd/common/blake3.d(34): Error: cannot implicitly convert expression `cursor + 1024LU` of type `ulong` to `uint`
```

**Comment by Geod24** [comment]
> Where is there an attempted implicit cast to uint?

It fails on x86, so when `.length` is 32 bits, not 64. Use `size_t` for `cursor` ?

File: compiler/src/dmd/common/blake3.d [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+ *
+ * Params:
+ *     data = byte array to hash
+ * Returns: Blake 3 hash of data
+ **/
+public ubyte[32] blake3(const ubyte[] data)
+{
+
+
+    ChunkState state;
```
Comment (RazvanN7) [inline compiler/src/dmd/common/blake3.d:22]: The 2 empty lines above do not match the dmd style. Please delete them.

File: compiler/src/dmd/common/blake3.d [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+    CVStack cvStack;
+
+    size_t cursor = 0;
```
Comment (RazvanN7) [inline compiler/src/dmd/common/blake3.d:25]: Also, I would group these 3 vars. There is no need for the empty line between them.

File: compiler/src/dmd/common/blake3.d [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+
+    //greater because if it's == we still need to finalize the last
+    //chunk
+    while(data.length - cursor > ChunkLength){
```
Comment (RazvanN7) [inline compiler/src/dmd/common/blake3.d:29]: `{` on their own line.

File: compiler/src/dmd/common/blake3.d [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+        state.block[] = 0;
+        state.blocksCompressed = 0;
+
+        cursor += ChunkLength;
+    }
+
+    //now handle the final chunk which might not be full
+
+    //handle all but last block
+    while(data.length - cursor > BlockLength){
```
Comment (RazvanN7) [inline compiler/src/dmd/common/blake3.d:48]: ditto

**Review by RazvanN7 — COMMENTED** [review]
I have only style comments:

- we typically use only one empty line to separate code paragraphs
- we use curly braces on their own lines
- variable declarations at the start of the function are placed immediately on the next line after the opening curly.

**Comment by RazvanN7** [comment]
cc @WalterBright

File: compiler/src/dmd/common/blake3.d [hunk 5]
```diff
@@ -0,0 +1,345 @@
+module dmd.common.blake3;
+// based on https://github.com/oconnor663/blake3_reference_impl_c/blob/main/reference_impl.c
+
+nothrow:
```
Comment (dkorpel) [inline compiler/src/dmd/common/blake3.d:4]: Can this be `@safe:` as well?

File: compiler/src/dmd/common/blake3.d [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+    uint[8] pop()
+    {
+        return data[--size];
+    }
+}
+
+
+
+void addChunkToTree(ref CVStack cvStack, uint[8] cv, ulong totalChunks){
```
Comment (dkorpel) [inline compiler/src/dmd/common/blake3.d:274]:
```suggestion
void addChunkToTree(ref CVStack cvStack, uint[8] cv, ulong totalChunks)
{
```

File: compiler/src/dmd/common/blake3.d [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        state[i + 8] ^= chainingValue[i];
+    }
+
+    return state;
+}
+
+
+//if block isn't full, only the first blockLength/4 words
+//will be filled in
+uint[16] bytesToWords(const ubyte[] block){
```
Comment (dkorpel) [inline compiler/src/dmd/common/blake3.d:209]:
```suggestion
uint[16] bytesToWords(const ubyte[] block)
{
```

File: compiler/src/dmd/common/blake3.d [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    g(state[3], state[7], state[11], state[15], message[6], message[7]);
+
+    //diagonals
+    g(state[0], state[5], state[10], state[15], message[8], message[9]);
+    g(state[1], state[6], state[11], state[12], message[10], message[11]);
+    g(state[2], state[7], state[8], state[13], message[12], message[13]);
+    g(state[3], state[4], state[9], state[14], message[14], message[15]);
+}
+
+void permute(ref uint[16] block){
```
Comment (dkorpel) [inline compiler/src/dmd/common/blake3.d:164]:
```suggestion
void permute(ref uint[16] block)
{
```

File: compiler/src/dmd/backend/cgobj.d [hunk 9]
```diff
@@ -2206,16 +2206,14 @@ size_t OmfObj_mangle(Symbol *s,char *dest)
...
         {
-            /* Form md5 digest of the name and store it in the
+            /* Form blake3 hash of the name and store it in the
              * last 32 bytes of the name.
              */
-            MD5_CTX mdContext;
-            MD5Init(&mdContext);
-            MD5Update(&mdContext, cast(ubyte *)name, cast(uint)len);
-            MD5Final(&mdContext);
+            const hash = blake3((cast(ubyte*)name)[0 .. len]);
```
Comment (dkorpel) [inline compiler/src/dmd/backend/cgobj.d:2212]:
```suggestion
            const hash = blake3((cast(ubyte*) name)[0 .. len]);
```

**Review by dkorpel — APPROVED** [review]
I'm curious about the benchmark result, though I expect it to not make a visible dent.

**Comment by benjones** [comment]
It's a busy week, so doubt I'll get a chance to do the experiment for a while.  I agree that I doubt there will be a measurable effect, but agree it would be good to confirm that.

**Comment by benjones** [comment]
OK, I ran a quick experiment.  Using master, compiled with dmd-2.107.0 and my branch compiled with dmd-2.107.0 I timed building phobos on a linux machine.  My branch took 21.33s at best,  master took 21.13s at best.  Not super thorough, but I think this shows that this change has negligible impact on build times.

**Comment by benjones** [comment]
ping

**Pull request merged**